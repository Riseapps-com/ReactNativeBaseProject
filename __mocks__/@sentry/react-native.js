module.exports = {
  ...jest.requireActual('@sentry/react-native'),

  init: jest.fn(),
  captureMessage: jest.fn(),
  captureException: jest.fn(),
};
