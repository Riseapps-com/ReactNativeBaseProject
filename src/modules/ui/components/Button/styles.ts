import { colors, sizes, styleSheetFactory } from '~theme';

export default styleSheetFactory(theme => ({
  button: {
    borderRadius: sizes.SPACING_S,
  },

  buttonContent: {
    height: sizes.SPACING_XXL + sizes.SPACING_S,
  },

  disabled: {
    backgroundColor: theme.disabled,
  },

  primary: {
    backgroundColor: theme.primary,
  },

  secondary: {
    backgroundColor: theme.accent,
  },

  textPrimary: {
    color: theme.secondaryText,
  },

  textSecondary: {
    color: theme.secondaryText,
  },

  textButton: {
    borderRadius: 0,
    backgroundColor: colors.TRANSPARENT,
  },
}));
