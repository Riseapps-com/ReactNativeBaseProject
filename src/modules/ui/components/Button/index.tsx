import React from 'react';
import { StyleProp, TextStyle, ViewStyle } from 'react-native';
import { Button as PaperButton } from 'react-native-paper';

import { useTheme } from '~theme';

import { ButtonTextType, ButtonType } from '../../types';
import themedStyles from './styles';

export type ButtonProps = Omit<
  React.ComponentProps<typeof PaperButton>,
  'dark' | 'disabled' | 'loading'
> & {
  buttonType?: ButtonType;
  style?: StyleProp<ViewStyle>;
  labelStyle?: StyleProp<TextStyle>;
  isLoading?: boolean;
  isDisabled?: boolean;
};

const DEFAULT_MODE = 'contained';
const DEFAULT_BUTTON_TYPE = 'primary';

const Button: React.FC<ButtonProps> = props => {
  const { mode, children, buttonType, labelStyle, style, isLoading, isDisabled, ...restProps } =
    props;
  const buttonMode = mode || DEFAULT_MODE;
  const defaultButtonType = buttonType || DEFAULT_BUTTON_TYPE;
  // eslint-disable-next-line
  const textStyleKey = `text${defaultButtonType.charAt(0).toUpperCase()}${defaultButtonType.slice(1)}` as ButtonTextType;
  const [styles] = useTheme(themedStyles);

  const buttonStyles: StyleProp<ViewStyle>[] = [
    styles.button,
    buttonMode === 'contained' && styles[defaultButtonType],
    buttonMode === 'text' && styles.textButton,
    isDisabled && styles.disabled,
    style,
  ];

  const labelStyles: StyleProp<TextStyle>[] = [styles[textStyleKey], labelStyle];

  return (
    <PaperButton
      style={buttonStyles}
      labelStyle={labelStyles}
      contentStyle={styles.buttonContent}
      mode={buttonMode}
      disabled={isDisabled}
      loading={isLoading}
      uppercase={false}
      {...restProps}
    >
      {children}
    </PaperButton>
  );
};

export default React.memo(Button);
