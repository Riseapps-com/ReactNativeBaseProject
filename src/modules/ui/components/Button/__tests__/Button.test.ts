import { fireEvent, RenderAPI } from '@testing-library/react-native';

import { renderComponent } from '~modules/tests';
import { light } from '~theme';

import Button, { ButtonProps } from '../index';

const children = 'Button text';
const onPress = jest.fn();

const defaultProps: ButtonProps = {
  children,
  onPress,
};

const renderButton = (props: Partial<ButtonProps> = {}) =>
  renderComponent(Button, { ...defaultProps, ...props }) as RenderAPI;

describe('ui', () => {
  describe('<Button />', () => {
    it('renders correctly', () => {
      const button = renderButton();

      expect(button).toMatchSnapshot();
    });

    it('renders with correct text', () => {
      const button = renderButton();

      expect(button.getByText(children)).toBeTruthy();
    });

    describe('press', () => {
      afterEach(() => {
        jest.clearAllMocks();
      });

      it('calls provided function', () => {
        const button = renderButton();

        fireEvent.press(button.getByText(children));

        expect(onPress).toBeCalledTimes(1);
      });
    });

    describe('button type', () => {
      it('renders "primary" button by default', () => {
        const button = renderButton();

        expect(button.toJSON()).toHaveStyle({ backgroundColor: light.primary });
      });

      it('renders "primary" button when specified', () => {
        const button = renderButton({ buttonType: 'primary' });

        expect(button.toJSON()).toHaveStyle({ backgroundColor: light.primary });
      });

      it('renders "secondary" button when specified', () => {
        const button = renderButton({ buttonType: 'secondary' });

        expect(button.toJSON()).toHaveStyle({
          backgroundColor: light.accent,
        });
      });

      it('renders "disabled" button when specified', () => {
        const button = renderButton({ isDisabled: true });

        expect(button.toJSON()).toHaveStyle({ backgroundColor: light.disabled });
      });

      it('overrides button style defaults', () => {
        const style = { backgroundColor: 'pink' };
        const customStyleButton = renderButton({ style });

        expect(customStyleButton.toJSON()).toHaveStyle(style);
      });
    });

    describe('text style', () => {
      it('renders light text for default button', () => {
        const button = renderButton();

        expect(button.getByText(children)).toHaveStyle({
          color: light.secondaryText,
        });
      });

      it('renders light text for "primary" button', () => {
        const button = renderButton({ buttonType: 'primary' });

        expect(button.getByText(children)).toHaveStyle({
          color: light.secondaryText,
        });
      });

      it('renders primary text for "secondary" button', () => {
        const button = renderButton({ buttonType: 'secondary' });

        expect(button.getByText(children)).toHaveStyle({
          color: light.secondaryText,
        });
      });

      it('renders disabled text from disabled button', () => {
        const button = renderButton({ isDisabled: true });

        expect(button.getByText(children)).toHaveStyle({ color: light.secondaryText });
      });

      it('overrides button text style defaults', () => {
        const labelStyle = { fontSize: 48, color: 'pink' };
        const customTextButton = renderButton({ labelStyle });

        expect(customTextButton.getByText(children)).toHaveStyle(labelStyle);
      });
    });

    describe('mode', () => {
      it('renders "contained" button by default', () => {
        const button = renderButton();

        expect(button.toJSON()).toHaveStyle({
          backgroundColor: light.primary,
          elevation: 2,
        });
      });

      it('renders specified button mode', () => {
        const textButton = renderButton({ mode: 'text' });

        expect(textButton.toJSON()).toHaveStyle({
          backgroundColor: 'transparent',
          elevation: 0,
        });
      });
    });
  });
});
