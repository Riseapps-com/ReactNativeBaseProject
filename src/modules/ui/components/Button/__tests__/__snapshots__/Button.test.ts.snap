// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui <Button /> renders correctly 1`] = `
<View
  collapsable={false}
  style={
    Object {
      "backgroundColor": "#0036A7",
      "borderColor": "transparent",
      "borderRadius": 8,
      "borderStyle": "solid",
      "borderWidth": 0,
      "elevation": 2,
      "minWidth": 64,
      "shadowColor": "#000000",
      "shadowOffset": Object {
        "height": 0.75,
        "width": 0,
      },
      "shadowOpacity": 0.24,
      "shadowRadius": 1.5,
    }
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": false,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Array [
        Object {
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 8,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "justifyContent": "center",
          },
          Object {
            "height": 48,
          },
        ]
      }
    >
      <Text
        numberOfLines={1}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#000000",
              "fontFamily": "System",
              "fontWeight": "400",
            },
            Object {
              "textAlign": "left",
            },
            Array [
              Object {
                "letterSpacing": 1,
                "marginHorizontal": 16,
                "marginVertical": 9,
                "textAlign": "center",
              },
              undefined,
              false,
              Object {
                "color": "#ffffff",
                "fontFamily": "System",
                "fontWeight": "500",
              },
              Object {
                "fontFamily": "System",
                "fontWeight": "500",
              },
              Array [
                Object {
                  "color": "#FFFFFF",
                },
                undefined,
              ],
            ],
          ]
        }
      >
        Button text
      </Text>
    </View>
  </View>
</View>
`;
